components/server/src/billing/cancel-chargebee-migration.ts

```diff
@@ -0,0 +1,191 @@
+import { AccountingDB, TeamSubscription2DB, TeamSubscriptionDB } from "@gitpod/gitpod-db/lib";
+import { ChargebeeService } from "../../ee/src/user/chargebee-service";
+import { Subscription } from "@gitpod/gitpod-protocol/lib/accounting-protocol";
+import { Plans } from "@gitpod/gitpod-protocol/lib/plans";
+import { log } from "@gitpod/gitpod-protocol/lib/util/logging";
+import { TeamSubscription, TeamSubscription2 } from "@gitpod/gitpod-protocol/lib/team-subscription-protocol";
+
+@injectable()
+export class CancelChargebeePersonalSubscriptionsMigration implements LongRunningMigration {
+    @inject(AccountingDB) protected readonly db: AccountingDB;
+    @inject(ChargebeeService) protected readonly chargebeeService: ChargebeeService;
+
+    getName(): string {
+        return "cancel-chargebee-personal-subscriptions";
+    }
+
+    /**
+     * This migration cancels all Chargebee Personal subscriptions in the database that are still active and not yet cancelled.
+     */
+    async runMigrationBatch(): Promise<boolean> {
+        const now = new Date().toISOString();
+
+        const subscriptions = await this.db.findActiveSubscriptions(now, now, 100);
+        // Same filter as in gitpod-server-impl:calculatePayAsYouGoNotifications
+        const activeNotCancelledSubscriptions = subscriptions.filter(
+            (s) => Plans.isPersonalPlan(s.planId) && !Plans.isFreePlan(s.planId) && !Subscription.isCancelled(s, now), // We only care about existing, active, not-yet-cancelled subs
+        );
+
+        let todo = activeNotCancelledSubscriptions.length;
+        for (const subscription of activeNotCancelledSubscriptions) {
+            const userId = subscription.userId;
+            if (!subscription.paymentReference) {
+                log.warn({ userId }, "Cancel: Subscription without payment reference", {
+                    subscriptionId: subscription.uid,
+                });
+                todo = -1;
+                continue;
+            }
+            if (subscription.paymentReference.startsWith("github:")) {
+                log.warn({ userId }, "Cancel: GitHub subscription", {
+                    subscriptionId: subscription.uid,
+                    paymentReference: subscription.paymentReference,
+                });
+                todo = -1;
+                continue;
+            }
+
+            try {
+                const chargebeeSubscriptionId = subscription.paymentReference;
+                await this.chargebeeService.cancelSubscription(
+                    chargebeeSubscriptionId,
+                    { userId },
+                    { subscriptionId: subscription.uid, paymentReference: subscription.paymentReference },
+                );
+                log.info({ userId }, "Cancel: Subscription cancelled", {
+                    subscriptionId: subscription.uid,
+                    paymentReference: subscription.paymentReference,
+                });
```

Comment:
I might be misunderstanding the intended use of `todo`, but should this case (where it worked) also decrement `todo`?

Reply:
same thought.